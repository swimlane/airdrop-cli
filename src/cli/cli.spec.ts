import { system } from 'gluegun/system';
import { join } from 'path';
import * as jetpack from 'fs-jetpack';

const colorsRe = /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g;
const doneRe = /\[.*s\]\n$/g;

const testDir = join(process.cwd(), '/test');
const testPath = join(testDir, '-/');
const binFile = join(process.cwd(), '/bin/airdrop');

describe('cli tests', () => {
  beforeAll(async () => {
    process.chdir(testDir);
    if (jetpack.exists(testPath)) {
      await jetpack.remove(testPath);
      await jetpack.remove(join(__dirname, 'airdrop.config.js'));
      return;
    }
  });

  describe('meta commands', () => {
    test('displays the version number', async () => {
      const version = await system.exec(`${binFile} --version`);
      expect(version.replace(colorsRe, '')).toBe('1.0.0');
    });

    test('displays help', async () => {
      const version = await system.exec(`${binFile} --help`);
      expect(version.replace(colorsRe, '')).toContain('airdrop version 1.0.0');
    });
  });

  describe('init', () => {
    let output;

    beforeAll(async () => {
      output = await system.exec(`${binFile} init`);
      output = output.replace(colorsRe, '').replace(doneRe, '');
    });

    test('displays console messages', async () => {
      expect(output).toMatchSnapshot();
    });

    test('files exist', async () => {
      expect(await jetpack.exists(join(testDir, 'airdrop.config.js'))).toBe('file');
    });

    test('can\'t init again', async () => {
      const out = await system.exec(`${binFile} init`);
      expect(out).toContain('airdrop.config.js already exists at');
      expect(out).toContain('skipping');
    });
  });

  describe('add', () => {
    let output;

    beforeAll(async () => {
      output = await system.exec(`${binFile} add lit-element@2.1.0`);
      output = output.replace(colorsRe, '').replace(doneRe, '');
    });

    test('displays console messages', async () => {
      expect(output).toMatchSnapshot();
    });

    test('files exist', async () => {
      expect(await jetpack.exists(join(testPath, 'lit-element@2.1.0'))).toBe('dir');
      expect(await jetpack.exists(join(testPath, 'lit-html@1.0.0'))).toBe('dir');
      expect(await jetpack.exists(join(testPath, 'importmap.json'))).toBe('file');
    });

    test('importmap', async () => {
      expect(await jetpack.readAsync(join(testPath, 'importmap.json'))).toMatchSnapshot();
    });

    test('can\t add again', async () => {
      const out = await system.exec(`${binFile} add lit-element@2.1.0`);
      expect(out).toContain('Package lit-element@2.1.0 already exists, skipping');
    });

    test('can force', async () => {
      let out = await system.exec(`${binFile} add lit-element@2.1.0 --force`);
      out = output.replace(colorsRe, '').replace(doneRe, '');
      expect(out).toMatchSnapshot();
    });
  });

  describe('resolve', () => {
    test('displays console messages', async () => {
      let output = await system.exec('../bin/airdrop resolve lit-element@2.1.0');
      output = output.replace(colorsRe, '').replace(doneRe, '');
      expect(output).toMatchSnapshot();
    });

    test('displays console messages', async () => {
      // Later this should fail if version is not installed
      let output = await system.exec(`${binFile} resolve lit-element@1.0.0`);
      output = output.replace(colorsRe, '').replace(doneRe, '');
      expect(output).toMatchSnapshot();
    });
  });

  describe('bundle', () => {
    let output;

    beforeAll(async () => {
      output = await system.exec(`${binFile} bundle lit-element@2.1.0`);
      output = output.replace(colorsRe, '').replace(doneRe, '');
    });

    test('displays console messages', async () => {
      expect(output).toMatchSnapshot();
    });

    test('files exist', async () => {
      expect(await jetpack.exists(join(testPath, 'lit-element@2.1.0.bundle.js'))).toBe('file');
    });

    test('can\'t bundle again', async () => {
      const out = await system.exec(`${binFile} bundle lit-element@2.1.0`);
      expect(out).toContain('Bundle already exists at');
      expect(out).toContain('skipping');
    });

    test('can force', async () => {
      let out = await system.exec(`${binFile} bundle lit-element@2.1.0 --force`);
      out = output.replace(colorsRe, '').replace(doneRe, '');
      expect(out).toMatchSnapshot();
    });
  });
});