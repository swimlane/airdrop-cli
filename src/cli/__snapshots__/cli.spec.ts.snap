// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli tests add can force 1`] = `
"Reading existing importmap
Fetching package information for lit-element@2.1.0
Fetching dependecy tree for lit-element@2.1.0
Extracting tarball for lit-element@2.1.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/lit-element@2.1.0
Extracting tarball for lit-html@1.0.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/lit-html@1.0.0
Writing importmap
Done! "
`;

exports[`cli tests add displays console messages 1`] = `
"Reading existing importmap
Fetching package information for lit-element@2.1.0
Fetching dependecy tree for lit-element@2.1.0
Extracting tarball for lit-element@2.1.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/lit-element@2.1.0
Extracting tarball for lit-html@1.0.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/lit-html@1.0.0
Writing importmap
Done! "
`;

exports[`cli tests add importmap 1`] = `
"{
  \\"imports\\": {
    \\"lit-element@2.1.0\\": \\"/-/lit-element@2.1.0/lit-element.js\\",
    \\"lit-element@2.1.0/\\": \\"/-/lit-element@2.1.0/\\"
  },
  \\"scopes\\": {
    \\"lit-element@2.1.0\\": {
      \\"lit-html\\": \\"/-/lit-html@1.0.0/lit-html.js\\",
      \\"lit-html/\\": \\"/-/lit-html@1.0.0/\\"
    },
    \\"lit-html@1.0.0\\": {}
  }
}"
`;

exports[`cli tests bundle can force 1`] = `
"Reading existing importmap
Fetching package information for d3@5.9.2
Fetching dependecy tree for d3@5.9.2
Extracting tarball for d3@5.9.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3@5.9.2
Extracting tarball for d3-zoom@1.7.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-zoom@1.7.3
Extracting tarball for d3-transition@1.2.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-transition@1.2.0
Extracting tarball for d3-timer@1.0.9 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-timer@1.0.9
Extracting tarball for d3-selection@1.4.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-selection@1.4.0
Extracting tarball for d3-interpolate@1.3.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-interpolate@1.3.2
Extracting tarball for d3-color@1.2.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-color@1.2.3
Extracting tarball for d3-ease@1.0.5 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-ease@1.0.5
Extracting tarball for d3-dispatch@1.0.5 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-dispatch@1.0.5
Extracting tarball for d3-drag@1.2.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-drag@1.2.3
Extracting tarball for d3-voronoi@1.1.4 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-voronoi@1.1.4
Extracting tarball for d3-time-format@2.1.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-time-format@2.1.3
Extracting tarball for d3-time@1.0.11 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-time@1.0.11
Extracting tarball for d3-shape@1.3.5 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-shape@1.3.5
Extracting tarball for d3-path@1.0.7 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-path@1.0.7
Extracting tarball for d3-scale-chromatic@1.3.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-scale-chromatic@1.3.3
Extracting tarball for d3-scale@2.2.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-scale@2.2.2
Extracting tarball for d3-format@1.3.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-format@1.3.2
Extracting tarball for d3-collection@1.0.7 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-collection@1.0.7
Extracting tarball for d3-array@1.2.4 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-array@1.2.4
Extracting tarball for d3-random@1.1.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-random@1.1.2
Extracting tarball for d3-quadtree@1.0.6 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-quadtree@1.0.6
Extracting tarball for d3-polygon@1.0.5 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-polygon@1.0.5
Extracting tarball for d3-hierarchy@1.1.8 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-hierarchy@1.1.8
Extracting tarball for d3-geo@1.11.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-geo@1.11.3
Extracting tarball for d3-force@1.2.1 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-force@1.2.1
Extracting tarball for d3-fetch@1.1.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-fetch@1.1.2
Extracting tarball for d3-dsv@1.1.1 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-dsv@1.1.1
Extracting tarball for rw@1.3.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/rw@1.3.3
Extracting tarball for iconv-lite@0.4.24 to /Users/jmh/workspace/projects/airdrop-cli/test/-/iconv-lite@0.4.24
Extracting tarball for safer-buffer@2.1.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/safer-buffer@2.1.2
Extracting tarball for commander@2.20.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/commander@2.20.0
Extracting tarball for d3-contour@1.3.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-contour@1.3.2
Extracting tarball for d3-chord@1.0.6 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-chord@1.0.6
Extracting tarball for d3-brush@1.0.6 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-brush@1.0.6
Extracting tarball for d3-axis@1.0.12 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-axis@1.0.12
Fetching package info for d3@5.9.2
Bundling d3@5.9.2
Writing bundle for d3@5.9.2
Writing importmap
Done! "
`;

exports[`cli tests bundle displays console messages 1`] = `
"Reading existing importmap
Fetching package information for d3@5.9.2
Fetching dependecy tree for d3@5.9.2
Extracting tarball for d3@5.9.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3@5.9.2
Extracting tarball for d3-zoom@1.7.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-zoom@1.7.3
Extracting tarball for d3-transition@1.2.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-transition@1.2.0
Extracting tarball for d3-timer@1.0.9 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-timer@1.0.9
Extracting tarball for d3-selection@1.4.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-selection@1.4.0
Extracting tarball for d3-interpolate@1.3.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-interpolate@1.3.2
Extracting tarball for d3-color@1.2.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-color@1.2.3
Extracting tarball for d3-ease@1.0.5 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-ease@1.0.5
Extracting tarball for d3-dispatch@1.0.5 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-dispatch@1.0.5
Extracting tarball for d3-drag@1.2.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-drag@1.2.3
Extracting tarball for d3-voronoi@1.1.4 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-voronoi@1.1.4
Extracting tarball for d3-time-format@2.1.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-time-format@2.1.3
Extracting tarball for d3-time@1.0.11 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-time@1.0.11
Extracting tarball for d3-shape@1.3.5 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-shape@1.3.5
Extracting tarball for d3-path@1.0.7 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-path@1.0.7
Extracting tarball for d3-scale-chromatic@1.3.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-scale-chromatic@1.3.3
Extracting tarball for d3-scale@2.2.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-scale@2.2.2
Extracting tarball for d3-format@1.3.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-format@1.3.2
Extracting tarball for d3-collection@1.0.7 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-collection@1.0.7
Extracting tarball for d3-array@1.2.4 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-array@1.2.4
Extracting tarball for d3-random@1.1.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-random@1.1.2
Extracting tarball for d3-quadtree@1.0.6 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-quadtree@1.0.6
Extracting tarball for d3-polygon@1.0.5 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-polygon@1.0.5
Extracting tarball for d3-hierarchy@1.1.8 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-hierarchy@1.1.8
Extracting tarball for d3-geo@1.11.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-geo@1.11.3
Extracting tarball for d3-force@1.2.1 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-force@1.2.1
Extracting tarball for d3-fetch@1.1.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-fetch@1.1.2
Extracting tarball for d3-dsv@1.1.1 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-dsv@1.1.1
Extracting tarball for rw@1.3.3 to /Users/jmh/workspace/projects/airdrop-cli/test/-/rw@1.3.3
Extracting tarball for iconv-lite@0.4.24 to /Users/jmh/workspace/projects/airdrop-cli/test/-/iconv-lite@0.4.24
Extracting tarball for safer-buffer@2.1.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/safer-buffer@2.1.2
Extracting tarball for commander@2.20.0 to /Users/jmh/workspace/projects/airdrop-cli/test/-/commander@2.20.0
Extracting tarball for d3-contour@1.3.2 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-contour@1.3.2
Extracting tarball for d3-chord@1.0.6 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-chord@1.0.6
Extracting tarball for d3-brush@1.0.6 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-brush@1.0.6
Extracting tarball for d3-axis@1.0.12 to /Users/jmh/workspace/projects/airdrop-cli/test/-/d3-axis@1.0.12
Fetching package info for d3@5.9.2
Bundling d3@5.9.2
Writing bundle for d3@5.9.2
Writing importmap
Done! "
`;

exports[`cli tests init displays console messages 1`] = `
"airdrop.config.js already exists at /Users/jmh/workspace/projects/airdrop-cli/test, skipping
Done! "
`;

exports[`cli tests resolve displays console messages 1`] = `"/-/lit-element@2.1.0/lit-element.js"`;

exports[`cli tests resolve displays console messages 2`] = `"/-/lit-element@1.0.0/lit-element.js"`;
